File: Codes/OptiBED/common/mhd_io.cpp
/*
 *  mhd_io.cpp
 *
 *  Fred project
 *
 *  Created by A.Schiavi 
 *  Copyright 2020. All rights reserved.
 *
 */


// purpose: minimal IO routines for mhd file format

#include "mhd_io.h"
#include "InputTools.h"
#include <iostream>
using namespace std;

string tolower(const string &str){
    string s = str;
    for(int i=0;i<s.length();i++) s[i] = ::tolower(s[i]);
    return s;
}

int icompare(std::string const& a, std::string const& b)
{
    if (a.length()<b.length()) return -1;
    if (a.length()>b.length()) return +1;
    for(int i=0;i<a.length();i++) if(std::tolower(a[i])!=std::tolower(b[i])) return std::tolower(a[i])<std::tolower(b[i]) ? -1 : +1;
    return 0;
}


int read_mhd_header(string fname,string &dtype,int nn[3],float x0[3],float L[3],float left[3],float up[3],float front[3],string &dataFile,size_t &dataFileOffset)
{
	const bool debug = false;

	// set default transform matrix = Identity
	for(int i=0;i<3;i++) {left[i]=up[i]=front[i]=0;}
	left[0]=up[1]=front[2]=1; // eye


    ifstream fin(fname);
    if(!fin){cerr<<"Error: could not open "<<fname<<endl; exit(-1);}
    string tok;
    int NDims;
    float hs[3],offset[3];
    while(1){
        if(fin.eof()) break;
        fin>>tok;
        if(!fin) break; // error
        if(icompare(tok,"ObjectType")==0){
            fin>>tok;
            if(tok!="=") {cerr<<"Error in reading mhd header"<<endl; return 1;}
            fin>>tok;
            if(icompare(tok,"Image")) {cerr<<"Error in reading mhd header"<<endl; return 1;}
            continue;
        }
        if(icompare(tok,"NDims")==0){
            fin>>tok;
            if(tok!="=") {cerr<<"Error in reading mhd header"<<endl; return 1;}
            fin>>NDims;
            if(debug) cout<<"Ndims = "<<NDims<<endl;
            continue;
        }
        if(icompare(tok,"DimSize")==0){
            fin>>tok;
            if(tok!="=") {cerr<<"Error in reading mhd header"<<endl; return 1;}
            if(debug) cout<<"nn = ";
            for(int i=0;i<NDims;i++) {fin>>nn[i];if(debug) cout<<nn[i]<<' ';}
            if(debug) cout<<endl;
            continue;
        }
        if(icompare(tok,"ElementType")==0){
            fin>>tok;
            if(tok!="=") {cerr<<"Error in reading mhd header"<<endl; return 1;}
            fin>>dtype;
            if(debug) cout<<"ElementType = "<<dtype<<endl;
            continue;
        }
        if(icompare(tok,"ElementSpacing")==0){
            fin>>tok;
            if(tok!="=") {cerr<<"Error in reading mhd header"<<endl; return 1;}
            if(debug) cout<<"hs = ";
            for(int i=0;i<NDims;i++) {fin>>hs[i];if(debug) cout<<hs[i]<<' ';}
            if(debug) cout<<endl;
            continue;
        }
        if(icompare(tok,"Offset")==0){
            fin>>tok;
            if(tok!="=") {cerr<<"Error in reading mhd header"<<endl; return 1;}
            if(debug) cout<<"offset = ";
            for(int i=0;i<NDims;i++) {fin>>offset[i];if(debug) cout<<offset[i]<<' ';}
            if(debug) cout<<endl;
            continue;
        }
        if(icompare(tok,"TransformMatrix")==0){
            fin>>tok;
            if(tok!="=") {cerr<<"Error in reading mhd header"<<endl; return 1;}
            if(debug) cout<<"TransformMatrix = ";
            for(int i=0;i<NDims;i++) {fin>>left[i];if(debug) cout<<left[i]<<' ';}
            for(int i=0;i<NDims;i++) {fin>>up[i];if(debug) cout<<up[i]<<' ';}
            for(int i=0;i<NDims;i++) {fin>>front[i];if(debug) cout<<front[i]<<' ';}
            if(debug) cout<<endl;
            continue;
        }
        if(icompare(tok,"BinaryData")==0){
            fin>>tok;
            if(tok!="=") {cerr<<"Error in reading mhd header"<<endl; return 1;}
            fin>>tok;
            if(icompare(tok,"true")){cerr<<"Error: expected BinaryData = True"<<endl;return 1;}
            continue;
        }
        if(icompare(tok,"BinaryDataByteOrderMSB")==0){
            fin>>tok;
            if(tok!="=") {cerr<<"Error in reading mhd header"<<endl; return 1;}
            fin>>tok;
            if(icompare(tok,"false")){cerr<<"Error: expected BinaryDataByteOrderMSB = False"<<endl;return 1;}
            continue;
        }
        if(icompare(tok,"CompressedData")==0){
            fin>>tok;
            if(tok!="=") {cerr<<"Error in reading mhd header"<<endl; return 1;}
            fin>>tok;
            if(icompare(tok,"false")){cerr<<"Error: expected CompressedData = False"<<endl;return 1;}
            continue;
        }
        if(icompare(tok,"ElementDataFile")==0){
            fin>>tok;
            if(tok!="=") {cerr<<"Error in reading mhd header"<<endl; return 1;}
            fin>>tok;
            if(icompare(tok,"LOCAL")){
                // get current directory and substitute header file with data file
                vector<string> pathtokens = strtokens(fname,"/");
                // cout<<"fname "<<fname<<endl;
                // for(int i=0;i<pathtokens.size();i++) cout<<i<<" "<<pathtokens[i]<<endl;
                dataFile="";
                if(fname[0]=='/') dataFile="/";
                for(int i=0;i<pathtokens.size()-1;i++) dataFile+=pathtokens[i]+"/";
                dataFile += tok;
                // cout<<"ElementDataFile "<<dataFile<<endl;
                // cout<<endl;
                dataFileOffset=0;
            } else {
                dataFile=fname; // is LOCAL => same file
                if(fin.peek()=='\r') fin.seekg(1,fin.cur);
                if(fin.peek()=='\n') fin.seekg(1,fin.cur);
                dataFileOffset = fin.tellg();                
            }
            break;
        }
        // cout<<tok<<endl;
        // if(tok.rfind("ElementDataFile")!=string::npos) break;
    }
    if(debug) cout<<"data offset = "<<dataFileOffset<<endl;
    fin.close();

	// from mhd to fred 
	x0[0]=offset[0]-0.5*hs[0]*left[0]-0.5*hs[1]*up[0]-0.5*hs[2]*front[0];
	x0[1]=offset[1]-0.5*hs[0]*left[1]-0.5*hs[1]*up[1]-0.5*hs[2]*front[1];
	x0[2]=offset[2]-0.5*hs[0]*left[2]-0.5*hs[1]*up[2]-0.5*hs[2]*front[2];

    if(debug) cout<<"x0 = "<<x0[0]<<' '<<x0[1]<<' '<<x0[2]<<endl;

	L[0]=nn[0]*hs[0];
	L[1]=nn[1]*hs[1];
	L[2]=nn[2]*hs[2];

	return 0;
}

int read_mhd_data(string fname,string &dtype,int nn[3],size_t &fileOffset,char **data)
{
    ifstream fin;
    fin.open(fname,ios_base::in|ios_base::binary);
    fin.seekg(fileOffset,fin.beg);
    size_t numVxl = 1UL*nn[0]*nn[1]*nn[2];
    size_t buffSize;
    if(dtype == "MET_CHAR") { buffSize = 1;}
    if(dtype == "MET_SHORT") { buffSize = 2;}
    if(dtype == "MET_INT") { buffSize = 4;}
    if(dtype == "MET_LONG") { buffSize = 4;}
    if(dtype == "MET_UCHAR") { buffSize = 1;}
    if(dtype == "MET_USHORT") { buffSize = 2;}
    if(dtype == "MET_UINT") { buffSize = 4;}
    if(dtype == "MET_ULONG") { buffSize = 4;}
    if(dtype == "MET_FLOAT") { buffSize = 4;}
    if(dtype == "MET_DOUBLE") { buffSize = 8;}
    buffSize *= numVxl;
    // cout<<"buffSize "<<buffSize<<endl;
    // cout<<"fileOffset "<<fileOffset<<endl;
    char *buff = new char[buffSize];
    fin.read(buff,buffSize);
    if(!fin) return 1;
    fin.close();
    *data = buff;
    return 0;
}


 
int write_mhd(string fname,string &dtype,char *data,int nn[3],float x0[3],float L[3],float left[3],float up[3],float front[3], bool localData){

    // check if extension is present
    string basename;
    if(fname.size()<4) {basename = fname;}
    else if(fname.substr(fname.size()-4)==".mhd"){
        basename=fname.substr(0,fname.size()-4);
    } else{
        basename=fname;
    }
    // cout<<"fname "<<fname<<endl;
    // cout<<"basename "<<basename<<endl;
    // cout<<"mhd "<<basename+".mhd"<<endl;
    // cout<<"raw "<<basename+".raw"<<endl;
    // exit(0);
    fname=basename+".mhd";
    ofstream fout(fname);
    if(!fout) return 1;

    fout<<"ObjectType = Image"<<endl;
    fout<<"NDims = 3"<<endl;
    fout<<"DimSize = "<<nn[0]<<' '<<nn[1]<<' '<<nn[2]<<endl;
    fout<<"BinaryData = True"<<endl;
    fout<<"BinaryDataByteOrderMSB = False"<<endl;
    fout<<"CompressedData = False"<<endl;
    // fout<<"AnatomicalOrientation = RAI"<<endl;
    
    fout<<"TransformMatrix = ";
    fout<<left [0]<<' '<<left [1]<<' '<<left [2]<<' ';
    fout<<up   [0]<<' '<<up   [1]<<' '<<up   [2]<<' ';
    fout<<front[0]<<' '<<front[1]<<' '<<front[2]<<' ';
    fout<<endl;
    
    float spacing[3],delta[3],offset[3];
    spacing[0]=L[0]/nn[0];
    spacing[1]=L[1]/nn[1];
    spacing[2]=L[2]/nn[2];

    delta[0]=0.5f*spacing[0];
    delta[1]=0.5f*spacing[1];
    delta[2]=0.5f*spacing[2];

    offset[0]=x0[0]+delta[0]*left[0]+delta[1]*up[0]+delta[2]*front[0];
    offset[1]=x0[1]+delta[0]*left[1]+delta[1]*up[1]+delta[2]*front[1];
    offset[2]=x0[2]+delta[0]*left[2]+delta[1]*up[2]+delta[2]*front[2];  
    
    fout<<"Offset = "<<offset[0]<<' '<<offset[1]<<' '<<offset[2]<<endl;
    fout<<"ElementSpacing = "<<spacing[0]<<' '<<spacing[1]<<' '<<spacing[2]<<endl;
    fout<<"ElementType = "<<dtype<<endl;
    if(localData){
        fout<<"ElementDataFile = LOCAL"<<endl;
    } else {
        fout<<"ElementDataFile = "<<basename+".raw"<<endl;
    }

    fout.close();
    if(!fout) return 1;

    size_t sz=0;
    if(dtype == "MET_CHAR") { sz = 1;}
    if(dtype == "MET_SHORT") { sz = 2;}
    if(dtype == "MET_INT") { sz = 4;}
    if(dtype == "MET_LONG") { sz = 4;}
    if(dtype == "MET_UCHAR") { sz = 1;}
    if(dtype == "MET_USHORT") { sz = 2;}
    if(dtype == "MET_UINT") { sz = 4;}
    if(dtype == "MET_ULONG") { sz = 4;}
    if(dtype == "MET_FLOAT") { sz = 4;}
    if(dtype == "MET_DOUBLE") { sz = 8;}

    size_t buffSize = sz*nn[0]*nn[1]*nn[2];

    if(localData){
        fout.open(fname,ios_base::out|ios_base::binary|ios_base::app);
    } else {
        fname = basename+".raw";
        fout.open(fname,ios_base::out|ios_base::binary);
    }
    
    if(!fout) return 1;
    fout.write(data,buffSize);
    if(!fout) return 1;
    fout.close();
    if(!fout) return 1;    
    return 0;
}

File: Scripts/PHOTONS_FLASH/mhd_io.py
#!/usr/bin/env python3
########################################################################
########################################################################
# Fred project
# read and write 3d maps in mhd format 
# Sep 2019 A.Schiavi and N. Krah
########################################################################
########################################################################
import numpy as np
import sys,os,re
########################################################################
MET_TYPES = ['MET_DOUBLE','MET_FLOAT','MET_INT','MET_LONG','MET_SHORT','MET_CHAR','MET_UCHAR','MET_UINT','MET_ULONG','MET_USHORT']
NUMPY_DTYPES = [np.float64,np.float32,np.int32,np.int64,np.int16,np.int8,np.uint8,np.uint32,np.uint64,np.uint16]
def mhd_read(fname):
    if fname[-4:] not in ['.mhd','.mha']:
        # print('fname',fname)
        if os.path.exists(fname+'.mhd'):
            fname = fname+'.mhd'
        elif os.path.exists(fname+'.mha'):
            fname = fname+'.mha'
        else:
            print('Error: file path extension is not mhd or mha')
            exit(-1)
    try:
        header = mhd_getHeader(fname)
        voxels=getMinimalMHD(fname,header)

        if getParam('BinaryData',header,1)!='True':
            print('BinaryData must be True');exit(1)

        if getParam('BinaryDataByteOrderMSB',header,1)!='False':
            print('BinaryDataByteOrderMSB must be False');exit(1)

        if getParam('CompressedData',header,1)!='False':
            print('CompressedData must be False');exit(1)

        ElementSpacing = getParam('ElementSpacing',header,3)
        if ElementSpacing==None:
            ElementSpacing=np.ones(3)
        else:
            try:
                ElementSpacing = np.array(list(map(float,ElementSpacing)))
            except:
                print('ElementSpacing is not a floating point list: ',ElementSpacing)

        TransformMatrix = getParam('TransformMatrix',header,9)
        if TransformMatrix==None:
            TransformMatrix=np.array([1,0,0,0,1,0,0,0,1])
        else:
            try:
                TransformMatrix = np.array(list(map(float,TransformMatrix)))
            except:
                print('TransformMatrix is not a floating point list: ',TransformMatrix)
        voxels['TransformMatrix']=TransformMatrix

        Offset = getParam('Offset',header,3)
        if Offset==None:
            Offset=ElementSpacing*0.5
        else:
            try:
                Offset = np.array(list(map(float,Offset)))
            except:
                print('Offset is not a floating point list: ',Offset)
        
        # from Offset to x0
        x0 = Offset
        Delta = ElementSpacing*0.5
        x0 -= Delta[0]*voxels['TransformMatrix'][0:3]
        x0 -= Delta[1]*voxels['TransformMatrix'][3:6]
        x0 -= Delta[2]*voxels['TransformMatrix'][6:9]

        voxels['hs']=ElementSpacing/10. # mm => cm
        voxels['x0']=x0/10. # mm => cm


        # CenterOfRotation = getParam('CenterOfRotation',header,3)
        # if CenterOfRotation==None:
        #     CenterOfRotation=np.zeros(3)
        # else:
        #     try:
        #         CenterOfRotation = np.array(list(map(float,CenterOfRotation)))
        #     except:
        #         print('CenterOfRotation is not a floating point list: ',CenterOfRotation)
        # voxels['CenterOfRotation']=CenterOfRotation

        return voxels
    except:
        print('IO error: cannot read mhd map from file:', fname)
    return None

########################################################################
def getParam(pname,lines,nvals):
    for l in lines:
        if l.strip()=='':
            continue
        tokens = l.split()
        # print(tokens)
        if len(tokens)<3 or tokens[1]!='=':
            print('syntax error in mhd file at line: ',l)
            exit(-1)
        if tokens[0]==pname:
            if nvals==1:
                return tokens[2]
            else:
                if len(tokens)!=2+nvals:
                    print('not enough values for parameter',pname)
                    exit(-1)
                return tokens[2:]
    return None
########################################################################
def mhd_getHeader(fname):
    header=[]
    try:
        fin = open(fname,'rb')
        while True:
            line = fin.readline().decode('utf-8').strip()
            # print(line)
            header.append(line)
            m = re.match(r'ElementDataFile\s*=\s*(.*)\s*$',line)
            if m:
                break
    except:
        print('IO error: cannot open input file or read the header', fname)
        return None    
    return header
########################################################################
def mhd_getData(fname,ElementDataFile):
    buffer=None
    if ElementDataFile=='LOCAL':
        try:
            fin = open(fname,'rb')
            while True:
                line = fin.readline().decode('utf-8').strip()
                # print(line)
                m = re.match(r'ElementDataFile\s*=\s*(.*)\s*$',line)
                if m:
                    if m.groups()[0]=='LOCAL':
                        buffer = fin.read()
                    break
        except:
            print('IO error: cannot open input file or read data from it', fname)
            return None
        return buffer
    else: # external file
        try:
            rawFilePath = os.path.join(os.path.dirname(fname),ElementDataFile)
            fin  = open(rawFilePath,'rb')
            buffer = fin.read()
        except:
            print('IO error: cannot open or read input file',ElementDataFile)
            return None
        return buffer
########################################################################
def getMinimalMHD(fname,lines):
    if getParam('ObjectType',lines,1)!='Image':
        print('Error: ObjectType != Image'); exit(-1)

    NDims = getParam('NDims',lines,1)
    if NDims==None:
        print('Error: NDims not found'); exit(-1)
    try:
        NDims = int(NDims)
    except:
        print('NDims is not an integer: ',NDims)
    
    DimSize = getParam('DimSize',lines,NDims)
    if DimSize==None:
        print('Error: DimSize not found'); exit(-1)
    try:
        DimSize = np.array(list(map(int,DimSize)))
    except:
        print('DimSize is not an integer list: ',DimSize)

    ElementType = getParam('ElementType',lines,1)
    if ElementType==None:
        print('Error: ElementType not found'); exit(-1)
    if ElementType not in MET_TYPES:
        print('ElementType ',ElementType,'is not one of',MET_TYPES)

    ElementDataFile = getParam('ElementDataFile',lines[-1:],1)
    if ElementDataFile==None:
        print('Error: ElementDataFile must be last tag of MetaImageHeader'); exit(-1)
    # print(ElementDataFile)

    voxels={}
    voxels['ElementDataFile']=ElementDataFile

    buffer = mhd_getData(fname,ElementDataFile)
    
    if buffer!=None:
        try:
            voxels['Map'] = np.frombuffer(buffer,dtype=NUMPY_DTYPES[MET_TYPES.index(ElementType)])
        except:
            print('IO error: cannot convert buffer to map')
    else:
        print('IO error: cannot open or read input file',ElementDataFile)
        exit(1)

    if voxels['Map'].size != np.prod(DimSize):
        print('Error: size of data in raw file does not match: read ',voxels['Map'].size,' elements instead of',np.prod(DimSize))
        exit(1)
    voxels['Map'] = np.reshape(voxels['Map'],DimSize,order='F')
    # voxels['Map'] = np.transpose(voxels['Map'])
    
    voxels['nn']=np.array(DimSize)
    return voxels


########################################################################
def mhd_write(fname,voxels):

    if len(fname)<5 or fname[-4:]!='.mhd': # add mhd extension if not present already
        fname=fname+'.mhd'

    Offset = voxels['x0'].copy()
    Delta = voxels['hs']/2.0
    Offset += Delta[0]*voxels['TransformMatrix'][0:3]
    Offset += Delta[1]*voxels['TransformMatrix'][3:6]
    Offset += Delta[2]*voxels['TransformMatrix'][6:9]
    # print('x0',voxels['x0'])
    # print('Offset',Offset)
    

    try:
        fout = open(fname,'w')
        print('ObjectType = Image',file=fout)
        print('NDims =',voxels['Map'].ndim,file=fout)
        print('DimSize =',' '.join(map(str,voxels['Map'].shape)),file=fout)
        print('ElementType =', MET_TYPES[NUMPY_DTYPES.index(voxels['Map'].dtype)],file=fout)
        print('ElementSpacing =',' '.join(map(str,voxels['hs']*10.)),file=fout)
        print('Offset =',' '.join(map(str,Offset*10.)),file=fout)
        print('BinaryData = True',file=fout)
        print('BinaryDataByteOrderMSB = False',file=fout)
        print('CompressedData = False',file=fout)
        
        # print('DimSize =',' '.join(map(str,voxels['Map'].shape)))

        for k in voxels.keys():
            if k in ['ElementDataFile','Map','nn','hs','x0']:
                continue
            print(k,'=',' '.join(map(str,voxels[k])),file=fout)
        # print('ElementDataFile =',voxels['ElementDataFile'],file=fout)
        print('ElementDataFile = LOCAL',file=fout)
        fout.close()
    except:
        print('IO error: cannot save mhd header to file:', fname)

    try:
        fout = open(fname,'ab')
        fout.write(voxels['Map'].tobytes(order='F'))
        fout.close()
    except:
        print('IO error: cannot save voxel map to file:', voxels['ElementDataFile'])

########################################################################
def packVoxels(nn,hs,x0,Map):
    voxels={}

    voxels['nn'] = nn.copy()
    voxels['hs'] = hs.copy()
    voxels['x0'] = x0.copy()
    voxels['Map']= Map.copy()
    voxels['TransformMatrix'] = np.array([1,0,0,0,1,0,0,0,1])

    return voxels
########################################################################
def unpackVoxels(voxels):
    return [voxels['nn'],voxels['hs'],voxels['x0'],voxels['Map']]
########################################################################
if __name__ == "__main__":
    if len(sys.argv)<2:
        print('usage: filename')
        exit(1)
    print('\n\nreading.....\n\n')
    voxels = mhd_read(sys.argv[1])
    print('dims=',voxels['nn'])
    print('spacing=',voxels['hs'])
    print('offset=',voxels['x0'])
    x0 = voxels['x0']
    hs = voxels['hs']
    nn = voxels['nn']
    print('x-spatial extent= [',x0[0],',',x0[0]+nn[0]*hs[0],'] => ',nn[0]*hs[0])
    print('y-spatial extent= [',x0[1],',',x0[1]+nn[1]*hs[1],'] => ',nn[1]*hs[1])
    print('z-spatial extent= [',x0[2],',',x0[2]+nn[2]*hs[2],'] => ',nn[2]*hs[2])
    print('datatype=',voxels['Map'].dtype )
    print('range=',np.min(voxels['Map']),np.max(voxels['Map']))
    print('sum=',np.sum(voxels['Map']))

    
    # Map32=Map.astype(np.float32)
    print('\n\nwriting.....\n\n')
    mhd_write('Map32.mhd',voxels)

